src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java

```diff
@@ -0,0 +1,142 @@
+                            throw new InterruptedException();
+                        }
+                        long sleep = Math.min(expiration, nextNotify) - System.currentTimeMillis();
+                        if (sleep > 0) {
+                            Thread.sleep(sleep);
+                        }
+                        // A random straw poll of users concluded that 3 minutes without any visible progress in the logs
+                        // is the point after which people believe that the process is dead.
+                        nextNotify += TimeUnit.SECONDS.toMillis(180);
+                        long now = System.currentTimeMillis();
+                        if (now < expiration) {
+                            GHRateLimit current = github.getRateLimit();
+                            if (current.remaining > rateLimit.remaining
+                                    || current.getResetDate().getTime() > rateLimit.getResetDate().getTime()) {
+                                listener.getLogger().println(GitHubConsoleNote.create(now,
+                                        "GitHub API Usage: The quota may have been refreshed earlier than expected, rechecking..."
+                                ));
+                                break;
+                            }
+                            listener.getLogger().println(GitHubConsoleNote.create(now, String.format(
+                                    "GitHub API Usage: Still sleeping, now only %s remaining.",
+                                    Util.getTimeSpanString(expiration - now)
+                            )));
+                        }
+                    }
+                }
+            }
+        }
+    },
+
+    ThrottleOnOver(Messages.ApiRateLimitChecker_ThrottleOnOver()) {
+        @Override
+        public void checkApiRateLimit(@NonNull TaskListener listener, GitHub github) throws IOException, InterruptedException {
+            boolean check = true;
+            while (check) {
+                GHRateLimit rateLimit = github.rateLimit();
+                // the buffer is how much we want to avoid using to cover unplanned over-use
+                int buffer = Math.max(15, rateLimit.limit / 20);
+                // the burst is how much we want to allow for speedier response outside of the throttle
+                if (rateLimit.remaining >= buffer) {
+                    break;
+                }
+                final long expiration = System.currentTimeMillis() + ENTROPY.nextInt(65536); // approx 1 min
+                listener.getLogger().println(GitHubConsoleNote.create(System.currentTimeMillis(), String.format(
+                        "GitHub API Usage: Current quota has %d remaining (%d over buffer). Next quota of %d due now. Sleeping for %s.",
+                        rateLimit.remaining, buffer - rateLimit.remaining, rateLimit.limit,
+                        Util.getTimeSpanString(expiration - System.currentTimeMillis())
+                )));
```

Comment:
I may not be reading this right, but I think you still need some form of the thread sleep code: 

https://github.com/jenkinsci/github-branch-source-plugin/pull/237/files#diff-01aac1fd9801e36fb1ffcaf09b9f3d43R71-R98

With out that, this will lock the thread up continuously checking the rate limit.

Reply:
You are totally right: I'm lost sleep code :(